File: src/main/java/no/nav/bidrag/dokument/service/PDFDokumentProcessor.java
package no.nav.bidrag.dokument.service;

import no.nav.bidrag.transport.dokument.DocumentProperties;
import org.apache.pdfbox.io.IOUtils;
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.pdmodel.PDPage;
import org.apache.pdfbox.pdmodel.PDPageContentStream;
import org.apache.pdfbox.pdmodel.PDPageContentStream.AppendMode;
import org.apache.pdfbox.pdmodel.common.PDRectangle;
import org.apache.pdfbox.util.Matrix;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.Optional;

public class PDFDokumentProcessor {
  private static final Logger LOGGER = LoggerFactory.getLogger(PDFDokumentProcessor.class);

  private PDDocument document;

  private byte[] originalDocument;

  private DocumentProperties documentProperties;

  public byte[] process(byte[] dokumentFil, DocumentProperties documentProperties){
    if (!documentProperties.shouldProcess()){return dokumentFil;}

    this.originalDocument = dokumentFil;
    this.documentProperties = documentProperties;
    ByteArrayOutputStream documentByteStream = new ByteArrayOutputStream();
    try (PDDocument document = PDDocument.load(dokumentFil)) {
      this.document = document;

      if (documentProperties.resizeToA4()){
        konverterAlleSiderTilA4();
      }

      if (documentProperties.optimizeForPrint()){
        optimaliserForDobbelsidigPrinting();
      }

      this.document.save(documentByteStream);
      this.document.close();
      return documentByteStream.toByteArray();
    } catch (Exception e) {
      LOGGER.error("Det skjedde en feil ved prossesering av PDF dokument", e);
      return dokumentFil;
    } finally {
      IOUtils.closeQuietly(documentByteStream);
    }
  }

  public void optimaliserForDobbelsidigPrinting(){
    if (documentHasOddNumberOfPages() && documentProperties.hasMoreThanOneDocument() && !documentProperties.isLastDocument()){
      LOGGER.debug("Dokumentet har oddetall antall sider. Legger til en blank side på slutten av dokumentet.");
      document.addPage(new PDPage(PDRectangle.A4));
    }
  }

  private boolean documentHasOddNumberOfPages(){
    return document.getNumberOfPages() % 2 != 0;
  }

  private boolean isPageSizeA4(PDPage pdPage){
    var a4PageMediaBox = PDRectangle.A4;
    var pageMediaBox = pdPage.getMediaBox();
    var hasSameHeightAndWidth = isSameWithMargin(pageMediaBox.getHeight(), a4PageMediaBox.getHeight(), 1F) && isSameWithMargin(pageMediaBox.getWidth(), a4PageMediaBox.getWidth(), 1F);
    var hasSameHeightAndWidthRotated = isSameWithMargin(pageMediaBox.getWidth(), a4PageMediaBox.getHeight(), 1F) && isSameWithMargin(pageMediaBox.getHeight(), a4PageMediaBox.getWidth(), 1F);
    return hasSameHeightAndWidth || hasSameHeightAndWidthRotated;
  }

  private void konverterAlleSiderTilA4() throws IOException {
    LOGGER.debug("Konverterer {} sider til A4 størrelse. Filstørrelse {}", document.getNumberOfPages(), bytesIntoHumanReadable(this.originalDocument.length));
    for (int pageNumber = 0; pageNumber < document.getNumberOfPages(); pageNumber++) {
      var page = document.getPage(pageNumber);
      updatePageRotationToVertical(page);
      if (!isPageSizeA4(page)) {
        convertPageToA4(page);
      }
    }
  }

  private void updatePageRotationToVertical(PDPage page){
    if (isVertical(page) && page.getRotation() != 0){
      page.setRotation(0);
    } else if (isHorizontal(page) && page.getRotation() == 0 && !isPageSizeA4(page)){
      page.setRotation(90);
    }
  }

  private boolean isHorizontal(PDPage page){
    return !isVertical(page);
  }
  /*
     En side skal roteres til å være vertikalt kun hvis siden er dimensjonert slik at høyden > bredden. Ellers skal det ignoreres
   */
  private boolean isVertical(PDPage page) {
      return Optional.ofNullable(page.getMediaBox()).map((mediaBox)->mediaBox.getHeight()>mediaBox.getWidth()).orElse(false);
  }

  private void convertPageToA4(PDPage page) throws IOException {

    Matrix matrix = new Matrix();
    float xScale = PDRectangle.A4.getWidth() / page.getMediaBox().getWidth();
    float yScale = PDRectangle.A4.getHeight() / page.getMediaBox().getHeight();
    matrix.scale(xScale, yScale);

    try (PDPageContentStream contentStream = new PDPageContentStream(document, page, AppendMode.PREPEND, false)) {
      contentStream.transform(matrix);
    }

    page.setMediaBox(PDRectangle.A4);
    page.setCropBox(PDRectangle.A4);
  }

  private boolean isSameWithMargin(Float val1, Float val2, Float margin){
    return Math.abs(val1-val2)<margin;
  }

  public static byte[] fileToByte(File file) throws IOException {
    FileInputStream inputStream = new FileInputStream(file);
    byte[] byteArray = new byte[(int)file.length()];
    inputStream.read(byteArray);
    inputStream.close();
    return byteArray;
  }

  public static String bytesIntoHumanReadable(long bytes) {
    long kilobyte = 1024;
    long megabyte = kilobyte * 1024;
    long gigabyte = megabyte * 1024;
    long terabyte = gigabyte * 1024;

    if ((bytes >= 0) && (bytes < kilobyte)) {
      return bytes + " B";

    } else if ((bytes >= kilobyte) && (bytes < megabyte)) {
      return (bytes / kilobyte) + " KB";

    } else if ((bytes >= megabyte) && (bytes < gigabyte)) {
      return (bytes / megabyte) + " MB";

    } else if ((bytes >= gigabyte) && (bytes < terabyte)) {
      return (bytes / gigabyte) + " GB";

    } else if (bytes >= terabyte) {
      return (bytes / terabyte) + " TB";

    } else {
      return bytes + " Bytes";
    }
  }
}

File: src/main/java/no/nav/bidrag/dokument/service/PDFDokumentProcessor.kt
package no.nav.bidrag.dokument.service

import mu.KotlinLogging
import no.nav.bidrag.transport.dokument.DocumentProperties
import org.apache.pdfbox.Loader
import org.apache.pdfbox.io.IOUtils
import org.apache.pdfbox.io.MemoryUsageSetting
import org.apache.pdfbox.io.RandomAccessReadBuffer
import org.apache.pdfbox.multipdf.PDFMergerUtility
import org.apache.pdfbox.pdfwriter.compress.CompressParameters
import org.apache.pdfbox.pdmodel.PDDocument
import org.apache.pdfbox.pdmodel.PDPage
import org.apache.pdfbox.pdmodel.PDPageContentStream
import org.apache.pdfbox.pdmodel.PDPageContentStream.AppendMode
import org.apache.pdfbox.pdmodel.common.PDRectangle
import org.apache.pdfbox.util.Matrix
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.util.*
import kotlin.math.abs

private val log = KotlinLogging.logger {}

class PDFDokumentProcessor {
    private var document: PDDocument? = null
    private lateinit var originalDocument: ByteArray
    private var documentProperties: DocumentProperties? = null
    fun process(
        dokumentFil: ByteArray,
        documentProperties: DocumentProperties,
        withCompression: Boolean = true
    ): ByteArray {
        if (!documentProperties.shouldProcess()) {
            return dokumentFil
        }
        originalDocument = dokumentFil
        this.documentProperties = documentProperties
        val documentByteStream = ByteArrayOutputStream()
        try {
            Loader.loadPDF(
                RandomAccessReadBuffer(dokumentFil),
                MemoryUsageSetting.setupTempFileOnly().streamCache,
            ).use { document ->
                this.document = document
                if (documentProperties.resizeToA4()) {
                    konverterAlleSiderTilA4()
                }
                if (documentProperties.optimizeForPrint()) {
                    optimaliserForDobbelsidigPrinting()
                }
                val compression =
                    if (withCompression) CompressParameters.DEFAULT_COMPRESSION else CompressParameters.NO_COMPRESSION
                this.document?.isAllSecurityToBeRemoved = true
                log.info { "Lagrer dokument med kompresjon ${compression.isCompress}" }
                this.document?.save(documentByteStream, compression)
                this.document?.close()
                return documentByteStream.toByteArray()
            }
        } catch (e: Exception) {
            log.error(e) { "Det skjedde en feil ved prossesering av PDF dokument" }
            return dokumentFil
        } catch (e: StackOverflowError) {
            log.error(e) { "Det skjedde en feil ved lagring av dokument med kompresjon. Forsøker å lagree dokument uten kompresjon" }
            return if (withCompression) process(dokumentFil, documentProperties, false)
            else dokumentFil
        } finally {
            IOUtils.closeQuietly(documentByteStream)
        }
    }

    fun optimaliserForDobbelsidigPrinting() {
        if (documentHasOddNumberOfPages() && documentProperties!!.hasMoreThanOneDocument() && !documentProperties!!.isLastDocument()) {
            log.debug("Dokumentet har oddetall antall sider. Legger til en blank side på slutten av dokumentet.")
            document?.addPage(PDPage(PDRectangle.A4))
        }
    }

    private fun documentHasOddNumberOfPages(): Boolean {
        return document != null && document!!.numberOfPages % 2 != 0
    }

    private fun isPageSizeA4(pdPage: PDPage): Boolean {
        val a4PageMediaBox = PDRectangle.A4
        val pageMediaBox = pdPage.mediaBox
        val hasSameHeightAndWidth =
            isSameWithMargin(pageMediaBox.height, a4PageMediaBox.height, 1f) && isSameWithMargin(
                pageMediaBox.width,
                a4PageMediaBox.width,
                1f,
            )
        val hasSameHeightAndWidthRotated =
            isSameWithMargin(pageMediaBox.width, a4PageMediaBox.height, 1f) && isSameWithMargin(
                pageMediaBox.height,
                a4PageMediaBox.width,
                1f,
            )
        return hasSameHeightAndWidth || hasSameHeightAndWidthRotated
    }

    @Throws(IOException::class)
    private fun konverterAlleSiderTilA4() {
        log.debug(
            "Konverterer {} sider til A4 størrelse. Filstørrelse {}",
            document!!.numberOfPages,
            bytesIntoHumanReadable(
                originalDocument.size.toLong(),
            ),
        )
        for (pageNumber in 0 until document!!.numberOfPages) {
            val page = document!!.getPage(pageNumber)
            updatePageRotationToVertical(page)
            if (!isPageSizeA4(page)) {
                convertPageToA4(page)
            }
        }
    }

    private fun updatePageRotationToVertical(page: PDPage) {
        if (isVertical(page) && page.rotation != 0) {
            page.rotation = 0
        } else if (isHorizontal(page) && page.rotation == 0 && !isPageSizeA4(page)) {
            page.rotation = 90
        }
    }

    private fun isHorizontal(page: PDPage): Boolean {
        return !isVertical(page)
    }

    /*
     En side skal roteres til å være vertikalt kun hvis siden er dimensjonert slik at høyden > bredden. Ellers skal det ignoreres
     */
    private fun isVertical(page: PDPage): Boolean {
        return Optional.ofNullable(page.mediaBox)
            .map { mediaBox: PDRectangle -> mediaBox.height > mediaBox.width }
            .orElse(false)
    }

    @Throws(IOException::class)
    private fun convertPageToA4(page: PDPage) {
        val matrix = Matrix()
        val xScale = PDRectangle.A4.width / page.mediaBox.width
        val yScale = PDRectangle.A4.height / page.mediaBox.height
        matrix.scale(xScale, yScale)
        PDPageContentStream(
            document,
            page,
            AppendMode.PREPEND,
            false,
        ).use { contentStream -> contentStream.transform(matrix) }
        page.mediaBox = PDRectangle.A4
        page.cropBox = PDRectangle.A4
    }

    private fun isSameWithMargin(val1: Float, val2: Float, margin: Float): Boolean {
        return abs(val1 - val2) < margin
    }

    companion object {

        @Throws(IOException::class)
        fun fileToByte(file: File): ByteArray {
            val inputStream = FileInputStream(file)
            val byteArray = ByteArray(file.length().toInt())
            inputStream.read(byteArray)
            inputStream.close()
            return byteArray
        }

        fun bytesIntoHumanReadable(bytes: Long): String {
            val kilobyte: Long = 1024
            val megabyte = kilobyte * 1024
            val gigabyte = megabyte * 1024
            val terabyte = gigabyte * 1024
            return if (bytes in 0..<kilobyte) {
                "$bytes B"
            } else if (bytes in kilobyte..<megabyte) {
                (bytes / kilobyte).toString() + " KB"
            } else if (bytes in megabyte..<gigabyte) {
                (bytes / megabyte).toString() + " MB"
            } else if (bytes in gigabyte..<terabyte) {
                (bytes / gigabyte).toString() + " GB"
            } else if (bytes >= terabyte) {
                (bytes / terabyte).toString() + " TB"
            } else {
                "$bytes Bytes"
            }
        }
    }
}

class PDFDokumentMerger {
    companion object {
        fun merge(
            dokumentBytes: List<ByteArray>,
            documentProperties: DocumentProperties,
            withCompression: Boolean = true
        ): ByteArray {
            documentProperties.numberOfDocuments = dokumentBytes.size
            if (dokumentBytes.size == 1) {
                return dokumentBytes[0]
            }
            val tempfiles = mutableListOf<File>()
            try {
                val mergedFileName = "/tmp/" + UUID.randomUUID()
                val mergedDocument = PDFMergerUtility()
                mergedDocument.destinationFileName = mergedFileName
                for (dokument in dokumentBytes) {
                    val tempFile = File.createTempFile("/tmp/" + UUID.randomUUID(), null)
                    tempFile.appendBytes(dokument)
                    tempfiles.add(tempFile)
                    mergedDocument.addSource(tempFile)
                }
                val compression = if (withCompression) CompressParameters.DEFAULT_COMPRESSION
                else CompressParameters.NO_COMPRESSION
                log.info { "Lagrer merget dokumenter med kompresjon ${compression.isCompress}" }
                mergedDocument.mergeDocuments(
                    MemoryUsageSetting.setupTempFileOnly().streamCache,
                    compression
                )
                return getByteDataAndDeleteFile(mergedFileName)
            } catch (e: StackOverflowError) {
                log.error(e) { "Det skjedde en feil ved merging av dokumenter. Forsøker å merge dokumenter uten kompresjon i legacy modus" }
                return mergeLegacy(dokumentBytes)
            } finally {
                tempfiles.forEach { it.delete() }
            }
        }

        private fun mergeLegacy(dokumentBytes: List<ByteArray>): ByteArray {
            if (dokumentBytes.size == 1) {
                return dokumentBytes[0]
            }
            val mergedDocument = PDDocument()
            val mergerUtility = PDFMergerUtility()
            val documentByteStream = ByteArrayOutputStream()
            log.info { "Merger ${dokumentBytes.size} dokumenter via legacy modus" }

            try {
                dokumentBytes.forEach {
                    Loader.loadPDF(
                        RandomAccessReadBuffer(it),
                        MemoryUsageSetting.setupTempFileOnly().streamCache,
                    ).use { document ->
                        mergerUtility.appendDocument(mergedDocument, document)
                        document.close()
                    }
                }

                mergedDocument.save(documentByteStream, CompressParameters.NO_COMPRESSION)
                return documentByteStream.toByteArray()
            } catch (e: StackOverflowError) {
                log.error(e) { "Det skjedde en feil ved merging av dokumenter i legacy modus." }
                throw e
            } catch (e: Exception) {
                log.error(e) { "Det skjedde en feil ved merging av dokumenter i legacy modus." }
                throw e
            } finally {
                IOUtils.closeQuietly(documentByteStream)
            }

        }

        private fun getByteDataAndDeleteFile(filename: String): ByteArray {
            val file = File(filename)
            return try {
                PDFDokumentProcessor.fileToByte(File(filename))
            } finally {
                file.delete()
            }
        }
    }

}

